exe/Main.hs

```diff
@@ -188,22 +192,256 @@ showEvent lock (EventFileDiagnostics (toNormalizedFilePath' -> file) diags) =
+        -- TODO Handle the case where there is no hie.yaml
+        -- Make a map from unit-id to DynFlags, this is used when trying to
+        -- resolve imports.
+        let uids = map (\(iuid, (df, _uis, _targets, _cfp, _di)) -> (iuid, df)) (new : old_deps)
+
+        -- For each component, now make a new HscEnvEq which contains the
+        -- HscEnv for the hie.yaml file but the DynFlags for that component
+        --
+        -- Then look at the targets for each component and create a map
+        -- from FilePath to the HscEnv
+        let new_cache (_iuid, (df, _uis, targets, cfp, di)) =  do
+              let hscEnv' = hscEnv { hsc_dflags = df
+                                   , hsc_IC = (hsc_IC hscEnv) { ic_dflags = df } }
+
+              versionMismatch <- evalGhcEnv hscEnv' checkGhcVersion
+              henv <- case versionMismatch of
+                        Just mismatch -> return mismatch
+                        Nothing -> newHscEnvEq hscEnv' uids
+              let res = (henv, di)
+              print res
+
+              let is = importPaths df
+              ctargets <- concatMapM (targetToFile is  . targetId) targets
+              -- A special target for the file which caused this wonderful
+              -- component to be created.
+              let special_target = (cfp, res)
+              --pprTraceM "TARGETS" (ppr (map (text . show) ctargets))
+              let xs = map (,res) ctargets
+              return (special_target:xs, res)
+
+        -- New HscEnv for the component in question
+        (cs, res) <- new_cache new
+        -- Modified cache targets for everything else in the hie.yaml file
+        -- which now uses the same EPS and so on
+        cached_targets <- concatMapM (fmap fst . new_cache) old_deps
+        modifyVar_ fileToFlags $ \var -> do
+            pure $ Map.insert hieYaml (HM.fromList (cs ++ cached_targets)) var
+
+        return res
+
+    lock <- newLock
+
+    -- This caches the mapping from hie.yaml + Mod.hs -> [String]
+    sessionOpts <- return $ \(hieYaml, file) -> do
+
+
+        fm <- readVar fileToFlags
+        let mv = Map.lookup hieYaml fm
+        let v = fromMaybe HM.empty mv
+        cfp <- liftIO $ canonicalizePath file
+        case HM.lookup (toNormalizedFilePath' cfp) v of
+          Just (_, old_di) -> do
+            deps_ok <- checkDependencyInfo old_di
+            unless deps_ok $ do
+              modifyVar_ fileToFlags (const (return Map.empty))
+              -- Keep the same name cache
+              modifyVar_ hscEnvs (return . Map.adjust (\(h, _) -> (h, [])) hieYaml )
+          Nothing -> return ()
+        -- We sort so exact matches come first.
+        case HM.lookup (toNormalizedFilePath' cfp) v of
+            Just opts -> do
+                --putStrLn $ "Cached component of " <> show file
+                pure (fst opts)
+            Nothing-> do
+                finished_barrier <- newBarrier
```

Comment:
I don’t understand this code. It sounds like you are trying to keep this running in the background so the next time you get here it will finish quickly if you get interrupted the first time. However, you will create a new barrier the next time you get here so it seems like the work of the thread is just wasted. Am I missing something?

Reply:
It seems you might be right that this `finished_barrier` is unnecessary now. I will try removing it.